dali/operators/generic/transpose/transpose.h

```diff
@@ -48,34 +48,53 @@ template <typename ShapeT = int>
 void PrepareArguments(SmallVector<ShapeT, kStaticShapeElements> &shape, VecInt &perm,
                       bool transpose = false) {
   DALI_ENFORCE(shape.size() == perm.size());
+  const int N = shape.size();
+
+  // This will be oterwise reduced to empty shape, and we still want to have some
+  // notion of non-empty shape left
+  if (volume(shape) == 1) {
+    shape = {1};
+    perm = {0};
+    return;
+  }
 
-  // cuTT does not handle dimensions with size 1 so we remove them
-  // (H, W, 1) is equivalent to (H, W)
-  auto it_shape = shape.begin();
-  auto it_perm = perm.begin();
-  SmallVector<ShapeT, kStaticShapeElements> erased;
-  while (it_shape != shape.end()) {
-    if (*it_shape == 1) {
-      erased.push_back(*it_perm);
-      it_shape = shape.erase(it_shape);
-      it_perm = perm.erase(it_perm);
+  SmallVector<int, kStaticShapeElements> ones_pos;
+  for (int i = 0; i < N; i++) {
+    if (shape[i] == 1) {
+      ones_pos.push_back(i);
+    }
+  }
+  SmallVector<ShapeT, kStaticShapeElements> tmp_shape;
+  tmp_shape.reserve(N - ones_pos.size());
+  // shape_idx holds index of original shape (already processed dimensions),
+  // ones_ids - holds indexes in array of degenerate (1-sized) dimensions
+  for (int shape_idx = 0, ones_idx = 0; shape_idx < N; shape_idx++) {
```

Comment:
Isn't this loop equivalent to this?
```
for (auto extent : shape)
  if (extent != 1)
    tmp_shape.push_back(extent);
```

Reply:
Yes it is, I initially wanted to take care of the permutation here as well. Will fix.